idl/subscribe.x

```diff
@@ -0,0 +1,131 @@
+namespace mazzaroth
+{
+  struct ReceiptSubscription
+  {
+    TransactionFilter transactionFilter;
+ 
+    ReceiptFilter receiptFilter;
+  };
+
+  struct ReceiptSubscriptionResult
+  {
+    Receipt receipt;
+
+    ID transactionID;
+  };
+
+  enum ValueFilterType
+  {
+    NONE = 0,
+    STRING = 1,
+    HASH32 = 2,
+    HASH64 = 3,
+    UHYPER = 4,
+    BOOL = 5,
+  };
+
+  union ValueFilter switch (ValueFilterType Type) 
+  {
+    case NONE:
+      void;
+    case STRING:
+      string regex<256>;
+    case HASH32:
+      Hash hash32Value;
+    case HASH64:
+      Signature hash64Value;
+    case UHYPER:
+      unsigned hyper uhyperValue;
+    case BOOL:
+      bool boolValue;
+  };
+
+  struct ReceiptValueFilter 
+  {
+    ValueFilter status;
+
+    ValueFilter stateRoot;
+  };
+
+  struct ActionFilter
```

Comment:
I feel like since every transaction has an Action it would make more sense to put the TransactionTypeFilter inside this struct and remove the ActionFilter from the individual types.  

Maybe rename this the TransactionFilter and the other one should be the CategoryFilter to better match the existing transaction structure?

Reply:
So you think it makes more sense to have the transaction filter split into 2 separate unions? I don't think this makes sense logically and involves double the amount of work whenever we want to construct or check a filter.

Reply:
I do think it makes more sense with the existing Transaction structure.  Transactions contain an Action, which has a category.  So I would expect the Transaction Filter to contain an Action Filter before specifying the Action Category or Transaction Type for specific filters.

This way you wouldn't need a new Transaction Type (GENERIC) if all you wanted to filter on were the Action fields.

I don't see how this would involve double the amount of work for constructing or checking filters.

Reply:
I don't understand what you're suggesting. Where would I put the ActionFilter? Inside which struct?

Reply:
So keeping the existing naming the same I would expect to see something like this:

```
struct ReceiptSubscription
  {
    ActionFilter actionFilter;
 
    ReceiptFilter receiptFilter;
  };

 struct ActionFilter
  {
    ValueFilter signature; 

    ValueFilter signer;

    ValueFilter address;

    ValueFilter channelID;

    ValueFilter nonce;

    TransactionFilter transactionFilter; // This would need to be renamed (something like CategoryFilter)
  };
```

And this would mean removing the ActionFilter from all of the other transaction type filters ( Contract, Config, Permission, Call) and Removing the Generic TransactionFilterType entirely.

Reply:
In this example you've added 5 unions that need to be checked to see if a transaction filter contains anything.
signature
signer
address
channelID
nonce

Maybe this makes the problem more clear. The json for an empty receipt subscription would go from:

{ transactionFilter: {enum:0, value:'""}, receiptFilter: {enum:0, value:""}}    ->
{ 
  actionFilter: {
    signature: {enum:0, value:'""},
    signer: {enum:0, value:'""},
    address: {enum:0, value:'""},
    channelid: {enum:0, value:'""},
    nonce: {enum:0, value:'""},
  },
  receiptFilter: {enum:0, value:""}}

Reply:
Okay, I see what the issue is now.  I was not aware that this was the reason for the high level transaction type union.  I still think it would be nice if there was a way to get rid of the need for a Generic transaction filter type, but if there is not I am okay with it being structured this way.